File: benchmarks/python/binarytrees.py
#!/usr/bin/env python3
"""Binary Trees benchmark aligned with examples/perf/binarytrees.flx."""

from __future__ import annotations

import sys

TIP = None


def make_tree(depth: int):
    if depth > 0:
        return (make_tree(depth - 1), make_tree(depth - 1))
    return (TIP, TIP)


def check_tree(tree) -> int:
    if tree is TIP:
        return 0
    left, right = tree
    return check_tree(left) + check_tree(right) + 1


def sum_trees(iterations: int, depth: int) -> int:
    total = 0
    for _ in range(iterations):
        total += check_tree(make_tree(depth))
    return total


def main() -> int:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 21
    min_depth = 4
    max_depth = max(min_depth + 2, n)
    stretch_depth = max_depth + 1

    print(
        f"stretch tree of depth {stretch_depth}\t check: "
        f"{check_tree(make_tree(stretch_depth))}"
    )

    long_lived_tree = make_tree(max_depth)

    for depth in range(min_depth, max_depth + 1, 2):
        iterations = 1 << (max_depth + min_depth - depth)
        total = sum_trees(iterations, depth)
        print(f"{iterations}\t trees of depth {depth}\t check: {total}")

    print(
        f"long lived tree of depth {max_depth}\t check: "
        f"{check_tree(long_lived_tree)}"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/rust/binarytrees.rs
use std::env;

#[derive(Clone)]
enum Tree {
    Node(Box<Tree>, Box<Tree>),
    Tip,
}

fn make_tree(depth: i32) -> Tree {
    if depth > 0 {
        Tree::Node(
            Box::new(make_tree(depth - 1)),
            Box::new(make_tree(depth - 1)),
        )
    } else {
        Tree::Node(Box::new(Tree::Tip), Box::new(Tree::Tip))
    }
}

fn check_tree(tree: &Tree) -> i32 {
    match tree {
        Tree::Node(left, right) => check_tree(left) + check_tree(right) + 1,
        Tree::Tip => 0,
    }
}

fn sum_trees(iterations: i32, depth: i32) -> i32 {
    let mut total = 0;
    for _ in 0..iterations {
        let tree = make_tree(depth);
        total += check_tree(&tree);
    }
    total
}

fn main() {
    let n = env::args()
        .nth(1)
        .and_then(|value| value.parse::<i32>().ok())
        .unwrap_or(21);
    let min_depth = 4;
    let max_depth = std::cmp::max(min_depth + 2, n);
    let stretch_depth = max_depth + 1;

    println!(
        "stretch tree of depth {}\t check: {}",
        stretch_depth,
        check_tree(&make_tree(stretch_depth))
    );

    let long_lived_tree = make_tree(max_depth);

    for depth in (min_depth..=max_depth).step_by(2) {
        let iterations = 1_i32 << (max_depth + min_depth - depth);
        let total = sum_trees(iterations, depth);
        println!(
            "{}\t trees of depth {}\t check: {}",
            iterations, depth, total
        );
    }

    println!(
        "long lived tree of depth {}\t check: {}",
        max_depth,
        check_tree(&long_lived_tree)
    );
}
